mcfly.zsh

```diff
@@ -0,0 +1,103 @@
+
+  # Populate McFly's temporary, per-session history file from recent commands in the shell's primary HISTFILE.
+  if [[ ! -f "${MCFLY_HISTORY}" ]]; then
+    export MCFLY_HISTORY=$(mktemp -t mcfly.XXXXXXXX)
+    tail -n100 "${HISTFILE}" >| ${MCFLY_HISTORY}
+  fi
+
+  if [[ $(fc -Iln 2>/dev/null | tail -n1) = ' mcfly search'* ]] ; then
+    # If the most recent history item is a mcfly search, do not update $MCFLY_HISTORY.
+    [ -n "$MCFLY_DEBUG" ] && echo "Not appending history to $MCFLY_HISTORY."
+  else
+    # Write history to $MCFLY_HISTORY.
+    fc -W "${MCFLY_HISTORY}"
+  fi
+
+  # Run mcfly with the saved code. It will:
+  # * append commands to $HISTFILE, (~/.zsh_history by default)
+  #   for backwards compatibility and to load in new terminal sessions;
+  # * find the text of the last command in $MCFLY_HISTORY and save it to the database.
+  [ -n "$MCFLY_DEBUG" ] && echo "Run mcfly add --exit ${exit_code}"
+  mcfly add --exit ${exit_code} # --append-to-histfile --zsh-extended-history
+  # Clear the in-memory history and reload it from $MCFLY_HISTORY
+  # (to remove instances of '#mcfly: ' from the local session history).
+  erase_history
+  fc -R "${MCFLY_HISTORY}"
+  return ${exit_code} # Restore the original exit code by returning it.
+}
+
+# Set $PROMPT_COMMAND run mcfly_prompt_command and then any existing $PROMPT_COMMAND.
+PROMPT_COMMAND="mcfly_prompt_command;$PROMPT_COMMAND"
+precmd_functions+=(mcfly_prompt_command)
+
+# Avoid logging #mcfly commands or commands starting with a space to the shell history.
+hist_filter() {
+  emulate -L zsh
+  local hist_cmd=${1%%$'\n'}
+  if [[ $hist_cmd = '#mcfly:'* ]] || [[ $hist_cmd = ' '* ]] ; then
+    [ -n "$MCFLY_DEBUG" ] && echo "Not saving: $hist_cmd"
+    return 2
+  else
+    [ -n "$MCFLY_DEBUG" ] && echo "Saving: $hist_cmd"
+    return 0
+  fi
+}
+zshaddhistory_functions+=(hist_filter)
+
+# Cleanup $MCFLY_HISTORY tmp files on exit.
+exit_logger() {
+  [ -n "$MCFLY_DEBUG" ] && echo "Exiting and removing $MCFLY_HISTORY"
+  rm -f $MCFLY_HISTORY
+}
+zshexit_functions+=(exit_logger)
+
+# If this is an interactive shell, take ownership of ctrl-r.
+# The logic here is:
+#   1. Jump to the beginning of the edit buffer, add 'mcfly: ', and comment out the current line. We comment out the line
+#      to ensure that all possible special characters, including backticks, are ignored. This commented out line will
+#      end up as the most recent entry in the $MCFLY_HISTORY file.
+#   2. Type "mcfly search" and then run the command. McFly will pull the last line from the $MCFLY_HISTORY file,
+#      which should be the commented-out search from step #1. It will then remove that line from the history file and
+#      render the search UI pre-filled with it.
+if [[ $- =~ .*i.* ]]; then
+  if set -o | grep "vi " | grep -q on; then
+    bindkey -s '\C-r' '\e0i#mcfly: \e\C-j mcfly search\C-j'
+  else
+    bindkey -s '\C-r' '\C-a#mcfly: \C-j mcfly search\C-j'
+  fi
+fi
```

Comment:
You can bind keybindings more robustly in zsh than bash - no need for translating keys or reading a file. Something like this should work, assuming you can call `mcfly search --query`:

```zsh
mcfly-history-widget() {
  local selected
  selected="$(mcfly search --query="${LBUFFER}")"
  local ret=$?
  if [[ -n $selected ]]; then
    LBUFFER="$selected"
  fi
  zle reset-prompt
  return $ret
}
zle -N mcfly-history-widget
bindkey '^R' fzf-history-widget
```

Reply:
Thanks, I'll try this out!

Reply:
@infokiller, here's a question for you.

`mcfly search` can be used in two ways:

1. Just running `mcfly search` brings up the UI and looks at the last line of `MCFLY_HISTORY`. If it matches "#mcfly: foo" then it prefills the UI with "foo". The rationale for this was that, at least in bash, I couldn't figure out how to safely pass the current command to mcfly while escaping all possible special characters.
2. `mcfly search "foo"` searches for "foo"

Do you know how I could use zsh's hooks to grab the current command buffer and pass it to mcfly in a fully-escaped way without using technique 1, which is what is done right now?

Thanks for your very helpful code review!

Reply:
Perhaps I don't understand something, but in the code example above, `mcfly search "${LBUFFER}"` should pass `"${LBUFFER}"` (which is what the user typed and is to the left of the cursor) with the exact string they typed. This uses standard shell escaping.
If however you want the argument passed to mcfly to be escaped when mcfly gets it, you can use `"${(q)var}"` (which quotes with backslashes), `"${(qq)var}"` (single quotes), `"${(qqq)var}"` (double quotes). Example output:

```zsh
❯ var='a $ " \\'
echo "${(q)var}"
echo "${(qq)var}"
echo "${(qqq)var}"
a\ \$\ \"\ \\
'a $ " \'
"a \$ \" \\"
```

Reply:
I think you're correct. I ran into a new problem, though. When I use your example, I get this error: `Os { code: 25, kind: Other, message: "Inappropriate ioctl for device" }`
It's coming out of the `selected="$(mcfly search --query="${LBUFFER}")"` call, and `stdout().into_raw_mode().unwrap()` in Rust. My guess is that I can't show a fullscreen UI when in this context.